### DISCUSSION TITLE
How can I debug the studio?

### DISCUSSION THREAD

**Discussion by ShuoShenDe** [foxglove/community#581] [Q&A]
Hello~Dear Foxglove group,

My name is Susan. I'm trying to debug the code of studio.
So I want to know, from which file the code start? 
I think the `desktop` folder is working for desktop server and desktop app, right? And the `web` folder is for web server, right? But I'm not sure what the other folders for.
Besides, when I try to set the breakpoints, they are always unbound. Could you tell me how to enable breakpoints?
![Screenshot from 2022-12-22 18-24-48](https://user-images.githubusercontent.com/31400000/209191510-8877edd7-b3bb-4d27-b066-e7f7997f85ff.png)

This is how I run it: I use the node.js server in VScode
![Screenshot from 2022-12-22 18-18-56](https://user-images.githubusercontent.com/31400000/209190635-70c181b6-c7b3-4b7e-aa90-02bea28c4db1.png)

**Comment by jtbandes** [comment]
I recommend setting breakpoints in the browser's dev tools.

**Reply by ShuoShenDe** [reply]
> I recommend setting breakpoints in the browser's dev tools.

But when you debug the desktop app, how you debug it?

**Reply by jtbandes** [reply]
Usually with breakpoints in the browser part of the app. It’s rare to need to debug the main process, I’m not sure I have any advice for it, but I’m sure you can find lots of guides for debugging Electron apps.

**Reply by ShuoShenDe** [reply]
> I recommend setting breakpoints in the browser's dev tools.

Hi~Dear jtbandes, but when I use the browser's dev tools, the code is very complex to read, and it is js, not ts. Just like the following. Am I missing something? Or is there any tutorial for browser dev tools? I want to debug on ts directly

![Screenshot from 2023-01-09 15-53-08](https://user-images.githubusercontent.com/31400000/211337204-4e28c928-9c48-4341-85fa-542a22489de6.png)

**Comment by defunctzombie** [comment]
As of 2.0 (https://foxglove.dev/blog/foxglove-2-0-unifying-robotics-observability) we no longer provide the source code for studio which also means we no longer provide support for working with the source code or debugging it.
